scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1515 @@
+                "Timestamp {} is out of the representable range for DateTime<Utc>",
+                millis
+            ))),
+        }
+    }
+);
+
+impl_strict_type!(
+    "time",
+    Time,
+    ColumnType::Time,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        let arr = ensure_exact_length::<8>("date", val)?;
+        let nanoseconds = i64::from_be_bytes(arr);
+
+        // Valid values are in the range 0 to 86399999999999
+        if !(0..=86399999999999).contains(&nanoseconds) {
+            return Err(ParseError::BadIncomingData(format!(
+                "Invalid time value only 0 to 86399999999999 allowed: {}.",
+                nanoseconds,
+            )));
+        }
+
+        Ok(Time(chrono::Duration::nanoseconds(nanoseconds)))
+    }
+);
+
+// inet
+
+impl_strict_type!(
+    "inet",
+    IpAddr,
+    ColumnType::Inet,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        if let Ok(ipv4) = <[u8; 4]>::try_from(val) {
+            Ok(IpAddr::from(ipv4))
+        } else if let Ok(ipv16) = <[u8; 16]>::try_from(val) {
+            Ok(IpAddr::from(ipv16))
+        } else {
+            Err(ParseError::BadIncomingData(format!(
+                "Invalid inet bytes length: {}",
+                val.len(),
+            )))
+        }
+    }
+);
+
+// uuid
+// TODO: Consider having separate types for timeuuid and uuid
+
+impl_strict_type!(
+    "timeuuid or uuid",
+    Uuid,
+    ColumnType::Uuid | ColumnType::Timeuuid,
+    |_typ: &'frame ColumnType, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null(v)?;
+        let arr = ensure_exact_length::<16>("timeuuid or uuid", val)?;
+        let i = u128::from_be_bytes(arr);
+        Ok(uuid::Uuid::from_u128(i))
+    }
+);
+
+/// A value that may be empty or not.
+///
+/// In CQL, some types can have a special value of "empty", represented as
+/// a serialized value of length 0. An example of this are integral types:
+/// the "int" type can actually hold 2^32 + 1 possible values because of this
+/// quirk. Note that this is distinct from being NULL.
+///
+/// `MaybeEmpty` was introduced to help support this quirk for Rust types
+/// which can't represent the empty, additional value.
+pub enum MaybeEmpty<T> {
+    Empty,
+    Value(T),
+}
+
+impl<'frame, T> DeserializeCql<'frame> for MaybeEmpty<T>
+where
+    T: DeserializeCql<'frame>,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        <T as DeserializeCql<'frame>>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        let val = ensure_not_null(v)?;
+        if val.is_empty() {
+            Ok(MaybeEmpty::Empty)
+        } else {
+            let v = <T as DeserializeCql<'frame>>::deserialize(typ, v)?;
+            Ok(MaybeEmpty::Value(v))
+        }
+    }
+}
```

Comment:
Hmm, I remember that I considered it when implementing `MaybeEmpty`. I think the reason for why I dropped was that it's easy to forget to add `EmptyableCqlType` when adding a new `impl DeserializeCql`. Besides, while `MaybeEmpty<String>` doesn't make much sense from the CQL protocol point of view, it does work and it should be pretty clear to the user what it does.

However, now that I think of it, it isn't a huge problem and adding missing `impl EmptyableCqlType` is easy and doesn't break the API. Moreover, unless users are well-versed with the CQL protocol spec (I wouldn't expect or require them to be), they won't know which types can be empty, so this might serve an informative purpose. In the future, if we rework the serialization API, then we'd rather use `MaybeEmpty` there, too - and the `EmptyableCqlType` will be even more important for the purpose of serialization.

To sum up - I'm convinced, I'll add a trait.